TimeOfFlight/include/TOFEstimators.h

```diff
@@ -29,69 +43,65 @@ class TH2F ;
+ *  For charged particles the TrackState at the calorimeter is used as reference point and for the
+ *  extrapolation into the calorimeter. For neutral particles the position of the first hit (the one closest to the IP)
  *  is used and a straight flight path from the IP is taken for the extrapolation.
  *  The hit time is corrected for the flight time from the reference assuming speed of light.
- *  
+ *
  *  See ../scripts/tofestimators.xml for example steering file.
  *
  * @author F.Gaede, DESY, April 2018
+ * @author B.Dudar, DESY, February 2021
  */
 
 class TOFEstimators : public Processor {
-  
- public:
-  
-  virtual Processor*  newProcessor() { return new TOFEstimators ; }
-  
-  TOFEstimators(const TOFEstimators&) = delete;
-  TOFEstimators& operator=(const TOFEstimators&) = delete;
-
-  TOFEstimators() ;
-  
-  /** Called at the begin of the job before anything is read.
-   * Use to initialize the processor, e.g. book histograms.
-   */
-  virtual void init() ;
-  
-  /** Called for every run.
-   */
-  virtual void processRunHeader( LCRunHeader* run ) ;
-  
-  /** Called for every event - the working horse.
-   */
-  virtual void processEvent( LCEvent * evt ) ; 
-  
-  
-  virtual void check( LCEvent * evt ) ; 
-  
-  
-  /** Called after data processing for clean up.
-   */
-  virtual void end() ;
-  
-  
- protected:
-
-  /** Input collection name with ReconstructedParticles (PFOs).
-   */
-  std::string _colNamePFO{};
-  int   _maxLayerNum{} ;
-  float _resolution{} ;
-  
-  std::vector<std::string> _TOFNames {} ; 
-
-  int _nRun{};
-  int _nEvt{};
-
-  gsl_rng* _rng = nullptr ;
-
-  std::vector<TH2F*> _h{};
-
-} ;
+    public:
 
-#endif
+    virtual Processor*  newProcessor() { return new TOFEstimators ; }
+
+    TOFEstimators(const TOFEstimators&) = delete;
+    TOFEstimators& operator=(const TOFEstimators&) = delete;
+
+    TOFEstimators() ;
 
+    /** Called at the begin of the job before anything is read.
+    * Use to initialize the processor, e.g. book histograms.
+    */
+    virtual void init() ;
 
+    /** Called for every event - the working horse.
+    */
+    virtual void processEvent(LCEvent* evt) ;
 
+    virtual void check(LCEvent* evt) ;
+
+    /** Called after data processing for clean up.
+    */
+    virtual void end() ;
+
+    Vector3D getMomAtCalo(const Track*);
+    double getFlightLength(const Track*);
+    double getTOFClosest(const Track*, const Cluster*);
+    double getTOFFastest(const Track*, const Cluster*);
+    double getTOFCylFit(const Track*, const Cluster*);
+    double getTOFClosestFit(const Track*, const Cluster*);
+
+    protected:
+
+    /** Input collection name with ReconstructedParticles (PFOs).
+    */
+    string _colNamePFO{};
+
+    string _procVersion{};
+    int _maxLayerNum{};
+    float _resolution{};
+    vector<string> _TOFNames{};
+
+    gsl_rng* _rng = nullptr;
+    vector<TH2F*> _h{};
+
+    std::default_random_engine _generator{};
```

Comment:
Using the std random number generator facilities is notoriously hard to do properly (see e.g. [here](https://codingnest.com/generating-random-numbers-using-c-standard-library-the-problems/) or [here](https://www.pcg-random.org/other-rngs.html#id28)), are also not guaranteed to be portable. Also the `std::default_random_engine` is implementation defined, so it might be better to use a dedicated generator here, to at least know which one is used.

Reply:
Hi Thomas,

I checked the links, they are very interesting and I agree. But I am not sure if this is so much of importance here. Also there is no such links for gsl generator, but I assume that is because of lack of checking it, rather than because it is perfect. So I would leave std one for the moment if you don't mind, otherwise switch to gsl is not a problem of course

Reply:
Yes, you are right, in this case it probably doesn't matter too much, but in general I think it is good to be aware of these potential problems. I am not sure about gsl either, but in general it should be OK, it just has a more C-like interface, which makes it a bit awkward to use in some cases.